File: 220. Contains Duplicate III/C++/Solution.cpp
class Solution {
public:
    bool containsNearbyAlmostDuplicate(vector<int> &nums, int k, int t) {
        if (k < 1 || t < 0) {
            return false;
        }
        unordered_map<long, long> buckets;
        for (int i = 0; i < nums.size(); i++) {
            long adjustedNum = (long) nums[i] - INT_MIN;
            long bucket = adjustedNum / (t + 1);
            if (buckets.find(bucket) != buckets.end()
                || (buckets.find(bucket - 1) != buckets.end() && abs(adjustedNum - buckets[bucket - 1]) <= t)
                || (buckets.find(bucket + 1) != buckets.end() && abs(adjustedNum - buckets[bucket + 1]) <= t)) {
                return true;
            }
            buckets[bucket] = adjustedNum;
            if (i >= k) {
                buckets.erase(((long) nums[i - k] - INT_MIN) / (t + 1));
            }
        }
        return false;
    }
};

File: 220. Contains Duplicate III/Java/Solution.java
public class Solution {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int k, int t) {
        if (k < 1 || t < 0) {
            return false;
        }
        Map<Long, Long> buckets = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            long adjustedNum = (long)nums[i] - Integer.MIN_VALUE;
            long bucket = adjustedNum / (t + 1);
            if (buckets.containsKey(bucket)
                || (buckets.containsKey(bucket - 1) && Math.abs(adjustedNum - buckets.get(bucket - 1)) <= t)
                || (buckets.containsKey(bucket + 1) && Math.abs(adjustedNum - buckets.get(bucket + 1)) <= t)) {
                return true;
            }
            buckets.put(bucket, adjustedNum);
            if (i >= k) {
                buckets.remove(((long) nums[i - k] - Integer.MIN_VALUE) / (t + 1));
            }
        }
        return false;
    }
}